File: src/main/java/com/dingdangmaoup/resin/pura/resin/jmx/ConnectorPingCommand.java
package com.dingdangmaoup.resin.pura.resin.jmx;

import com.dingdangmaoup.resin.pura.ResinModelBase;
import com.dingdangmaoup.resin.pura.resin.configuration.Resin3XConfigurationStrategy;
import java.io.IOException;
import javax.management.JMException;
import javax.management.MBeanServerConnection;

public class ConnectorPingCommand extends ConnectorCommandBase<Boolean> {

  private final int myJmxPort;

  public ConnectorPingCommand(ResinModelBase resinModel, int jmxPort) {
    super(resinModel);
    myJmxPort = jmxPort;
  }

  @Override
  protected int getJmxPort() {
    return myJmxPort;
  }

  // TODO: check behavior if ping is insuccessful - ex should not be logged
  @Override
  protected Boolean doExecute(MBeanServerConnection connection) throws JMException, IOException {
    // TODO: use server version - like DM: getServerVersion().getRecoveryMonitorMBean()
    return Resin3XConfigurationStrategy.STATE_JMX_ATTRIBUTE_ACTIVE.equalsIgnoreCase(
      (String)connection.getAttribute(Resin3XConfigurationStrategy.MBEAN_WEB_APP_DEPLOY, Resin3XConfigurationStrategy.STATE_JMX_ATTRIBUTE));
  }
}

File: src/main/kotlin/com/dingdangmaoup/resin/pura/resin/jmx/ConnectorPingCommand.kt
package com.dingdangmaoup.resin.pura.resin.jmx

import com.dingdangmaoup.resin.pura.ResinRemoteModel
import java.io.IOException
import javax.management.JMException
import javax.management.MBeanServerConnection

internal class ConnectorPingCommand(
    resinModel: ResinRemoteModel,
    private val endpoint: JmxEndpoint,
    credentialSource: JmxCredentialSource,
) : ConnectorCommandBase<Boolean>(
    resinModel,
    { resolveCredentialSource(credentialSource) { resinModel.loadJmxCredentials(endpoint) } },
) {
    override fun getHost(): String = endpoint.connectionHost

    override fun getJmxPort(): Int = endpoint.port

    @Throws(JMException::class, IOException::class)
    override fun doExecute(connection: MBeanServerConnection): Boolean {
        val state = connection.getAttribute(
            JmxDeploymentClient.MBEAN_WEB_APP_DEPLOY,
            JmxDeploymentClient.STATE_JMX_ATTRIBUTE,
        ) as String
        return JmxDeploymentClient.STATE_JMX_ATTRIBUTE_ACTIVE.equals(state, ignoreCase = true)
    }
}
